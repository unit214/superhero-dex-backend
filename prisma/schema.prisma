// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "debian-openssl-1.1.x"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Token {
  id         Int     @id @default(autoincrement())
  address    String  @unique
  symbol     String  @default("n/a")
  name       String  @default("n/a")
  decimals   Int     @default(-1)
  malformed  Boolean @default(false)
  noContract Boolean @default(false)
  pairs0     Pair[]  @relation("Pair_Token0")
  pairs1     Pair[]  @relation("Pair_Token1")
  listed     Boolean @default(false)
}

model Pair {
  id                          Int                               @id @default(autoincrement())
  address                     String                            @unique
  token0                      Token                             @relation("Pair_Token0", fields: [t0], references: [id])
  t0                          Int
  token1                      Token                             @relation("Pair_Token1", fields: [t1], references: [id])
  t1                          Int
  liquidityInfo               PairLiquidityInfo?                @relation("Pair_Info")
  synchronized                Boolean
  liquidityInfoHistory        PairLiquidityInfoHistory[]
  liquidityInfoHistoryV2      PairLiquidityInfoHistoryV2[]
  liquidityInfoHistoryError   PairLiquidityInfoHistoryError[]
  liquidityInfoHistoryV2Error PairLiquidityInfoHistoryV2Error[]
}

model PairLiquidityInfo {
  id          Int    @id
  pair        Pair   @relation("Pair_Info", fields: [id], references: [id])
  totalSupply String
  reserve0    String
  reserve1    String
  height      Int
}

model PairLiquidityInfoHistory {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  totalSupply    String
  reserve0       String
  reserve1       String
  height         Int
  microBlockHash String
  microBlockTime BigInt
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashUniqueIndex", [pairId, microBlockHash])
}

model PairLiquidityInfoHistoryError {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  microBlockHash String
  error          String
  timesOccurred  Int      @default(1)
  createdAt      DateTime @default(now())
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashErrorUniqueIndex", [pairId, microBlockHash, error])
}

model PairLiquidityInfoHistoryV2 {
  id              Int      @id @default(autoincrement())
  pair            Pair     @relation(fields: [pairId], references: [id])
  pairId          Int
  eventType       String
  height          Int
  microBlockTime  BigInt
  logIndex        Int
  microBlockHash  String
  transactionHash String
  totalSupply     String?
  reserve0        String?
  reserve1        String?
  deltaReserve0   String?
  deltaReserve1   String?
  fiatPrice       String?
  createdAt       DateTime @default(now())
  updatedAt       DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashLogIndexUniqueIndex", [pairId, microBlockHash, logIndex])
}

model PairLiquidityInfoHistoryV2Error {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  microBlockHash String
  logIndex       Int
  error          String
  timesOccurred  Int      @default(1)
  createdAt      DateTime @default(now())
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashLogIndexUniqueIndex", [pairId, microBlockHash, logIndex, error])
}
