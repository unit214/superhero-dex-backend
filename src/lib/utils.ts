import { Decimal } from '@prisma/client/runtime/library';
import BigNumber from 'bignumber.js';

export const nonNullable = <T>(t: T | null | undefined, label?: string): T => {
  if (t == null) {
    throw new Error(
      `${label ? label + ': ' : null} non nullable value expected`,
    );
  }
  return t;
};

export const removeId = <ID, T extends { id: ID }>(t: T) => {
  const {
    id, // eslint-disable-line @typescript-eslint/no-unused-vars
    ...tail
  } = t;
  return tail;
};

export const pluralize = (count: number, noun: string, suffix = 's') =>
  `${count} ${noun}${count !== 1 ? suffix : ''}`;

const parseEnv = (x) => x && JSON.parse(x);
export const presentInvalidTokens = parseEnv(process.env.SHOW_INVALID_TOKENS);

export const numberToDecimal = (number: number): Decimal =>
  new Decimal(number.toString());

export const bigIntToDecimal = (bigInt: bigint): Decimal =>
  new Decimal(bigInt.toString());

export const decimalToBigInt = (decimal: Decimal): bigint =>
  BigInt(decimal.toFixed().toString());

export const calculateUsdValue = ({
  reserve,
  tokenAePrice,
  decimals,
  aeUsdPrice,
}: {
  reserve: string;
  tokenAePrice: string;
  decimals: number;
  aeUsdPrice: string;
}) => {
  if (!reserve || !tokenAePrice || !aeUsdPrice) {
    return '0';
  }

  return new BigNumber(reserve)
    .abs()
    .div(new BigNumber(10).pow(decimals))
    .multipliedBy(tokenAePrice)
    .multipliedBy(aeUsdPrice)
    .toString();
};
