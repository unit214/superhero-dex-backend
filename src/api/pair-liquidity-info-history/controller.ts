import {
  Controller,
  Get,
  ParseEnumPipe,
  ParseIntPipe,
  Query,
} from '@nestjs/common';
import { PairLiquidityInfoHistoryService } from './service';
import { ApiOperation, ApiQuery, ApiResponse } from '@nestjs/swagger';
import * as dto from '../../dto';
import { OrderQueryEnum } from '../../dto';
import { ContractAddress } from '../../lib/utils';

@Controller('history/liquidity')
export class PairLiquidityInfoHistoryController {
  constructor(
    private readonly pairLiquidityInfoHistoryService: PairLiquidityInfoHistoryService,
  ) {}

  @Get()
  @ApiOperation({
    summary: 'Retrieve all entries of the pair liquidity info history',
    deprecated: true,
  })
  @ApiQuery({
    name: 'limit',
    type: Number,
    description: 'Limit of history entries per page (default: 100, max: 100)',
    required: false,
  })
  @ApiQuery({
    name: 'offset',
    type: Number,
    description: 'Offset of page (default: 0)',
    required: false,
  })
  @ApiQuery({
    name: 'order',
    enum: OrderQueryEnum,
    description:
      'Sorts history entries in ascending or descending order (default: asc)',
    required: false,
  })
  @ApiQuery({
    name: 'pair-address',
    type: String,
    description: 'Retrieve only history entries for the given pair address',
    required: false,
  })
  @ApiQuery({
    name: 'height',
    type: Number,
    description: 'Retrieve only history entries for the given height',
    required: false,
  })
  @ApiQuery({
    name: 'from-block-time',
    type: Number,
    description:
      'Retrieve only history entries that are equal or newer than the given micro block time',
    required: false,
  })
  @ApiQuery({
    name: 'to-block-time',
    type: Number,
    description:
      'Retrieve only history entries that are equal or older than the given micro block time',
    required: false,
  })
  @ApiResponse({ status: 200, type: [dto.PairLiquidityInfoHistoryEntry] })
  async findAll(
    @Query('limit', new ParseIntPipe({ optional: true })) limit: number = 100,
    @Query('offset', new ParseIntPipe({ optional: true })) offset: number = 0,
    @Query('order', new ParseEnumPipe(OrderQueryEnum, { optional: true }))
    order: OrderQueryEnum = OrderQueryEnum.asc,
    @Query('pair-address') pairAddress?: ContractAddress,
    @Query('height', new ParseIntPipe({ optional: true })) height?: number,
    @Query('from-block-time', new ParseIntPipe({ optional: true }))
    fromBlockTime?: number,
    @Query('to-block-time', new ParseIntPipe({ optional: true }))
    toBlockTime?: number,
  ): Promise<dto.PairLiquidityInfoHistoryEntry[]> {
    return this.pairLiquidityInfoHistoryService
      .getAllHistoryEntries(
        Number(limit),
        Number(offset),
        order,
        pairAddress,
        height ? Number(height) : undefined,
        fromBlockTime ? BigInt(fromBlockTime) : undefined,
        toBlockTime ? BigInt(toBlockTime) : undefined,
      )
      .then((entries) =>
        entries.map((entry) => ({
          pairAddress: entry.pair.address,
          liquidityInfo: {
            totalSupply: entry.totalSupply,
            reserve0: entry.reserve0,
            reserve1: entry.reserve1,
          },
          height: entry.height,
          microBlockHash: entry.microBlockHash,
          microBlockTime: entry.microBlockTime.toString(),
        })),
      );
  }
}
