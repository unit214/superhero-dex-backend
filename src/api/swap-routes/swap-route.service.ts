import { Injectable } from '@nestjs/common';

import { PairDbService } from '@/database/pair/pair-db.service';
import { presentInvalidTokens } from '@/lib/utils';

@Injectable()
export class SwapRoutesService {
  constructor(private readonly pairDbService: PairDbService) {}

  async getAllPairsWithLiquidityInfo(onlyListed?: boolean) {
    return this.pairDbService.getAllWithLiquidityInfo(
      presentInvalidTokens,
      onlyListed,
    );
  }
}
