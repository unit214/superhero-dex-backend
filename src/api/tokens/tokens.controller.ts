import {
  Controller,
  Get,
  Param,
  Headers,
  NotFoundException,
  Post,
  Delete,
  UnauthorizedException,
} from '@nestjs/common';
import { TokensService } from './tokens.service';
import * as dto from '../../dto';
import {
  ApiResponse,
  ApiOperation,
  ApiParam,
  ApiHeaders,
} from '@nestjs/swagger';
import { removeId, ContractAddress } from '../../lib/utils';
import * as prisma from '@prisma/client';

const withTokenAuthorization = async <T>(
  auth: string,
  work: () => Promise<T>,
): Promise<T> => {
  if (auth !== process.env.AUTH_TOKEN) {
    throw new UnauthorizedException('wrong authorization');
  }
  try {
    return await work(); // await is not redundant is for the try catch purpose
  } catch (error) {
    if (error.message.indexOf('Record to update not found') > -1) {
      throw new NotFoundException('Token not found');
    } else {
      throw error;
    }
  }
};
const toDtoToken = ({
  listed, // eslint-disable-line @typescript-eslint/no-unused-vars
  id, // eslint-disable-line @typescript-eslint/no-unused-vars
  ...tail
}: prisma.Token) => tail;
@Controller('tokens')
export class TokensController {
  constructor(private readonly tokensService: TokensService) {}

  @Get()
  @ApiOperation({
    summary: 'Retrieves all stored tokens',
    description: `All the tokens no matter if are officially supported by the DEX (listed=true) or not will be retrieved`,
  })
  @ApiResponse({ status: 200, type: [dto.TokenWithListed] })
  async getAllTokens(): Promise<dto.TokenWithListed[]> {
    return (await this.tokensService.getAllTokens()).map((token) =>
      removeId(token),
    );
  }

  @Get('listed')
  @ApiOperation({
    summary: 'Retrieves all officially supported tokens',
    description: `Just direct token information is retrieved, for more information as pair addresses or even full pair details
use \`tokens/:address\` or \`tokens/:address/pairs\` `,
  })
  @ApiResponse({ status: 200, type: [dto.Token] })
  async getListedTokens(): Promise<dto.Token[]> {
    return (await this.tokensService.getListedTokens()).map(toDtoToken);
  }

  @Get('by-address/:address')
  @ApiParam({
    name: 'address',
    required: true,
    example:
      process.env.DOC_TOKEN1 ||
      'ct_b7FZHQzBcAW4r43ECWpV3qQJMQJp5BxkZUGNKrqqLyjVRN3SC',
    ...dto.pairAddressPropertyOptions,
  })
  @ApiOperation({
    summary: 'Gets a specific token',
    description: `Gets information about a specific token with all belonging pair addresses`,
  })
  @ApiResponse({ status: 200, type: dto.TokenWithPairAddresses })
  @ApiResponse({ status: 404 })
  async findOne(
    @Param('address') address: ContractAddress,
  ): Promise<dto.TokenWithPairAddresses> {
    const token = await this.tokensService.getToken(address);
    if (!token) {
      throw new NotFoundException('token not found');
    }

    const {
      pairs0,
      pairs1,
      id, // eslint-disable-line @typescript-eslint/no-unused-vars
      ...tail
    } = token;
    return {
      ...tail,
      pairs: pairs0.concat(pairs1).map((x) => x.address),
    };
  }

  @Get('by-address/:address/pairs')
  @ApiParam({
    name: 'address',
    required: true,
    example:
      process.env.DOC_TOKEN1 ||
      'ct_b7FZHQzBcAW4r43ECWpV3qQJMQJp5BxkZUGNKrqqLyjVRN3SC',
    ...dto.pairAddressPropertyOptions,
  })
  @ApiOperation({
    summary: 'Retrieves all belonging pairs for a given token',
    description: `Gets all pairs with the liquidity attached and also the other pair token information. 
The returning object is splitted in two, pairs in which the given token represents the token0 and another list in which
the given token represents the token1`,
  })
  @ApiResponse({ status: 200, type: dto.TokenPairs })
  @ApiResponse({ status: 404 })
  async pairs(
    @Param('address') address: ContractAddress,
  ): Promise<dto.TokenPairs> {
    const token = await this.tokensService.getTokenWithPairsInfo(address);
    if (!token) {
      throw new NotFoundException('token not found');
    }

    return {
      pairs0: token.pairs0.map(
        ({ address, synchronized, liquidityInfo, token1 }) => ({
          address,
          synchronized,
          oppositeToken: removeId(token1),
          liquidityInfo:
            (liquidityInfo && removeId(liquidityInfo)) || undefined,
        }),
      ),
      pairs1: token.pairs1.map(
        ({ address, synchronized, liquidityInfo, token0 }) => ({
          address,
          synchronized,
          oppositeToken: removeId(token0),
          liquidityInfo:
            (liquidityInfo && removeId(liquidityInfo)) || undefined,
        }),
      ),
    };
  }

  @Post('listed/:address')
  @ApiParam({
    name: 'address',
    required: true,
    example:
      process.env.DOC_TOKEN2 ||
      'ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF',
    ...dto.pairAddressPropertyOptions,
  })
  @ApiOperation({
    summary: 'Mark token as listed',
    description: `Add a token in the official token list`,
  })
  @ApiHeaders([
    {
      name: 'Authorization',
      description: 'authorization token',
      required: true,
      allowEmptyValue: false,
      example:
        '77e1c923ec679c11a3a2efbd0cde5927edf874818e5be54443da6b12a9280202',
    },
  ])
  @ApiResponse({ status: 201, type: dto.TokenWithListed })
  @ApiResponse({ status: 404 })
  @ApiResponse({ status: 401 })
  async listToken(
    @Param('address') address: ContractAddress,
    @Headers('Authorization') auth: string,
  ): Promise<dto.TokenWithListed> {
    return withTokenAuthorization(auth, async () =>
      removeId(await this.tokensService.listToken(address)),
    );
  }

  @Delete('listed/:address')
  @ApiParam({
    name: 'address',
    required: true,
    example:
      process.env.DOC_TOKEN2 ||
      'ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF',
    ...dto.pairAddressPropertyOptions,
  })
  @ApiOperation({
    summary: 'Unmark token as listed',
    description: `Remove a token from the official token list`,
  })
  @ApiHeaders([
    {
      name: 'Authorization',
      description: 'authorization token',
      required: true,
      allowEmptyValue: false,
      example:
        '77e1c923ec679c11a3a2efbd0cde5927edf874818e5be54443da6b12a9280202',
    },
  ])
  @ApiResponse({ status: 200, type: dto.TokenWithListed })
  @ApiResponse({ status: 404 })
  @ApiResponse({ status: 401 })
  async unlistToken(
    @Param('address') address: ContractAddress,
    @Headers('Authorization') auth: string,
  ): Promise<dto.TokenWithListed> {
    return withTokenAuthorization(auth, async () =>
      removeId(await this.tokensService.unlistToken(address)),
    );
  }
}
