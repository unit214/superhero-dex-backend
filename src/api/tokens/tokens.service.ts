import { Injectable } from '@nestjs/common';
import * as dal from '../../dal';
import { Pair, Token, PairLiquidityInfo } from '@prisma/client';
import { ContractAddress, presentInvalidTokens } from '../../lib/utils';

@Injectable()
export class TokensService {
  async getCount(onlyListed?: boolean) {
    return dal.token.count(presentInvalidTokens, onlyListed);
  }
  async getAllTokens(): Promise<Token[]> {
    return dal.token.getAll(presentInvalidTokens);
  }
  async getListedTokens(): Promise<Token[]> {
    return dal.token.getListed();
  }

  async listToken(address: ContractAddress) {
    return dal.token.updateListedValue(address, true);
  }

  async unlistToken(address: ContractAddress) {
    return dal.token.updateListedValue(address, false);
  }

  async getToken(address: ContractAddress): Promise<
    | (Token & {
        pairs0: Pair[];
        pairs1: Pair[];
      })
    | null
  > {
    return dal.token.getByAddressWithPairs(address);
  }

  async getTokenWithPairsInfo(address: ContractAddress): Promise<
    | (Token & {
        pairs0: (Pair & {
          token1: Token;
          liquidityInfo: PairLiquidityInfo | null;
        })[];
        pairs1: (Pair & {
          liquidityInfo: PairLiquidityInfo | null;
          token0: Token;
        })[];
      })
    | null
  > {
    return dal.token.getByAddressWithPairsAndLiquidity(address);
  }
}
