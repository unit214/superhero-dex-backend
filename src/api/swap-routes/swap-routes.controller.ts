import { CacheInterceptor } from '@nestjs/cache-manager';
import { Controller, Get, Param, Query, UseInterceptors } from '@nestjs/common';
import { ApiOperation, ApiParam, ApiQuery, ApiResponse } from '@nestjs/swagger';
import * as prisma from '@prisma/client';

import * as dto from '@/api/api.model';
import { toLiquidityInfoDto } from '@/api/pairs/pairs.controller';
import { SwapRoutesService } from '@/api/swap-routes/swap-route.service';
import { getPaths } from '@/lib/paths';

const toPair = (
  pair: prisma.Pair & {
    token0: prisma.Token;
    token1: prisma.Token;
  },
) => {
  return {
    address: pair.address,
    token0: pair.token0.address,
    token1: pair.token1.address,
    synchronized: pair.synchronized,
  };
};

const toPairWithLiquidityInfo = (
  pair: prisma.Pair & {
    liquidityInfo: prisma.PairLiquidityInfo | null;
    token0: prisma.Token;
    token1: prisma.Token;
  },
): dto.PairWithLiquidityAndTokenAddresses => {
  return {
    ...toPair(pair),
    liquidityInfo: toLiquidityInfoDto(pair.liquidityInfo),
  };
};

@UseInterceptors(CacheInterceptor)
@Controller('swap-routes')
export class SwapRoutesController {
  constructor(private readonly swapRoutesService: SwapRoutesService) {}

  @ApiOperation({
    summary: 'Retrieves all swap routes between two tokens',
    description: `Gets all swap routes directly or with one intermediate token from one token to another. \
The routes are represented as arrays of pairs.<br\> <b>NOTE</b>: the response is an array of pair arrays, but \
because of nest-swagger limitation the documentation can't show a multi-dimensional array. Though, please take in consideration \
the real response type is \`Array<Array<PairWithLiquidityAndTokenAddresses>>\``,
  })
  @Get(':from/:to?')
  @ApiQuery({
    name: 'only-listed',
    type: Boolean,
    description:
      'Retrieves only the pairs having both tokens added in the official token list',
    required: false,
  })
  @ApiResponse({
    status: 200,
    type: dto.PairWithLiquidityAndTokenAddresses,
    isArray: true,
  })
  @ApiParam({
    name: 'from',
    required: true,
    example:
      process.env.DOC_TOKEN1 ||
      'ct_b7FZHQzBcAW4r43ECWpV3qQJMQJp5BxkZUGNKrqqLyjVRN3SC',
    ...dto.tokenAddressPropertyOptions,
  })
  @ApiParam({
    name: 'to',
    required: true,
    example:
      process.env.DOC_TOKEN2 ||
      'ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF',
    ...dto.tokenAddressPropertyOptions,
  })
  async getSwapRoutes(
    @Param('from') from: string,
    @Param('to') to: string,
    @Query('only-listed') onlyListedStr: string, //false | true
  ): Promise<dto.PairWithLiquidityAndTokenAddresses[][]> {
    const onlyListed = !!onlyListedStr && onlyListedStr !== 'false';
    const pairs =
      await this.swapRoutesService.getAllPairsWithLiquidityInfo(!!onlyListed);
    const edges = pairs.map((data) => ({
      data,
      t0: data.token0.address,
      t1: data.token1.address,
    }));
    return getPaths(from, to, edges).map((pairs) =>
      pairs.map((pair) => toPairWithLiquidityInfo(pair)),
    );
  }
}
