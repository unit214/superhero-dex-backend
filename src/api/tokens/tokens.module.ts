import { Module } from '@nestjs/common';
import { TokensController } from './tokens.controller';
import { TokensService } from './tokens.service';

@Module({
  imports: [],
  controllers: [TokensController],
  providers: [TokensService],
})
export class TokensModule {}
