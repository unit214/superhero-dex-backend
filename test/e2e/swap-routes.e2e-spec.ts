import { CacheModule } from '@nestjs/cache-manager';
import { INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import * as request from 'supertest';

import { SwapRoutesService } from '@/api/swap-routes/swap-route.service';
import { SwapRoutesController } from '@/api/swap-routes/swap-routes.controller';
import { TokensController } from '@/api/tokens/tokens.controller';
import { TokensService } from '@/api/tokens/tokens.service';
import { PairDbService } from '@/database/pair/pair-db.service';
import { PrismaService } from '@/database/prisma.service';
import { TokenDbService } from '@/database/token/token-db.service';
import { nonNullable } from '@/lib/utils';
import {
  historyEntry1,
  historyEntry2,
  historyEntry3,
  historyEntry4,
  liquidityInfo1,
  liquidityInfo2,
  pair1,
  pair2,
  pair3,
  pair4,
  token1,
  token2,
  token3,
  token4,
  token5,
} from '@/test/mock-data/pair-liquidity-info-history-mock-data';

describe('SwapRoutesController', () => {
  let app: INestApplication;
  let prismaService: PrismaService;

  const authToken = nonNullable(process.env.AUTH_TOKEN);

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [SwapRoutesController, TokensController],
      imports: [CacheModule.register({})],
      providers: [
        SwapRoutesService,
        PairDbService,
        PrismaService,
        TokensService,
        TokenDbService,
      ],
    }).compile();

    app = module.createNestApplication();
    await app.init();

    prismaService = module.get(PrismaService);
  });

  beforeAll(async () => {
    await prismaService.pairLiquidityInfoHistory.deleteMany();
    await prismaService.pairLiquidityInfo.deleteMany();
    await prismaService.pair.deleteMany();
    await prismaService.token.deleteMany();
  });

  beforeEach(async () => {
    await prismaService.token.createMany({
      data: [token1, token2, token3, token4, token5],
    });
    await prismaService.pair.createMany({ data: [pair1, pair2, pair3, pair4] });
    await prismaService.pairLiquidityInfo.createMany({
      data: [liquidityInfo1, liquidityInfo2],
    });
    await prismaService.pairLiquidityInfoHistory.createMany({
      data: [historyEntry1, historyEntry2, historyEntry3, historyEntry4],
    });
  });

  afterEach(async () => {
    await prismaService.pairLiquidityInfoHistory.deleteMany();
    await prismaService.pairLiquidityInfo.deleteMany();
    await prismaService.pair.deleteMany();
    await prismaService.token.deleteMany();
  });

  afterAll(async () => {
    await prismaService.$disconnect();
    await app.close();
  });

  describe('GET /swap-routes/{from}/{to}', () => {
    it('should return 200 with [] if no path for unexisting token ', async () => {
      return request(app.getHttpServer())
        .get('/swap-routes/ct_token1/ct_xxxx')
        .expect(200)
        .expect([]);
    });

    it('should return 200 with [] for existing token if no pair or path exists', async () => {
      return request(app.getHttpServer())
        .get('/swap-routes/ct_token1/ct_token4')
        .expect(200)
        .expect([]);
    });

    it('should return a direct path', async () => {
      return request(app.getHttpServer())
        .get('/swap-routes/ct_token2/ct_token3')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });

    it('should return an indirect path', async () => {
      return request(app.getHttpServer())
        .get('/swap-routes/ct_token1/ct_token3')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });

    it('should return one direct path and one indirect path', async () => {
      return request(app.getHttpServer())
        .get('/swap-routes/ct_token1/ct_token5')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });

    it('should suppress some paths with only-listed=true ', async () => {
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_token1')
        .set('Authorization', authToken);
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_token5')
        .set('Authorization', authToken);

      return request(app.getHttpServer())
        .get('/swap-routes/ct_token1/ct_token5?only-listed=true')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });

    it('should return paths oven on reverse order of tokens', async () => {
      return request(app.getHttpServer())
        .get('/swap-routes/ct_token1/ct_token5')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });
  });
});
