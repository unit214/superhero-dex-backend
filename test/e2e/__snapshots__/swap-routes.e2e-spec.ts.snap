// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwapRoutesController GET /swap-routes/{from}/{to} should return a direct path 1`] = `
[
  [
    {
      "address": "ct_pair2",
      "liquidityInfo": {
        "height": 1,
        "reserve0": "1",
        "reserve1": "1",
        "totalSupply": "1",
      },
      "synchronized": true,
      "token0": "ct_token2",
      "token1": "ct_token3",
    },
  ],
]
`;

exports[`SwapRoutesController GET /swap-routes/{from}/{to} should return an indirect path 1`] = `
[
  [
    {
      "address": "ct_pair1",
      "synchronized": true,
      "token0": "ct_token1",
      "token1": "ct_token2",
    },
    {
      "address": "ct_pair2",
      "liquidityInfo": {
        "height": 1,
        "reserve0": "1",
        "reserve1": "1",
        "totalSupply": "1",
      },
      "synchronized": true,
      "token0": "ct_token2",
      "token1": "ct_token3",
    },
  ],
]
`;

exports[`SwapRoutesController GET /swap-routes/{from}/{to} should return one direct path and one indirect path 1`] = `
[
  [
    {
      "address": "ct_pair4",
      "synchronized": true,
      "token0": "ct_token1",
      "token1": "ct_token5",
    },
  ],
  [
    {
      "address": "ct_pair1",
      "synchronized": true,
      "token0": "ct_token1",
      "token1": "ct_token2",
    },
    {
      "address": "ct_pair3",
      "liquidityInfo": {
        "height": 2,
        "reserve0": "2",
        "reserve1": "2",
        "totalSupply": "2",
      },
      "synchronized": true,
      "token0": "ct_token2",
      "token1": "ct_token5",
    },
  ],
]
`;

exports[`SwapRoutesController GET /swap-routes/{from}/{to} should return paths oven on reverse order of tokens 1`] = `
[
  [
    {
      "address": "ct_pair4",
      "synchronized": true,
      "token0": "ct_token1",
      "token1": "ct_token5",
    },
  ],
  [
    {
      "address": "ct_pair1",
      "synchronized": true,
      "token0": "ct_token1",
      "token1": "ct_token2",
    },
    {
      "address": "ct_pair3",
      "liquidityInfo": {
        "height": 2,
        "reserve0": "2",
        "reserve1": "2",
        "totalSupply": "2",
      },
      "synchronized": true,
      "token0": "ct_token2",
      "token1": "ct_token5",
    },
  ],
]
`;

exports[`SwapRoutesController GET /swap-routes/{from}/{to} should suppress some paths with only-listed=true  1`] = `
[
  [
    {
      "address": "ct_pair4",
      "synchronized": true,
      "token0": "ct_token1",
      "token1": "ct_token5",
    },
  ],
]
`;
